Provide better impacts for fields indexed with IndexOptions.DOCS

expani opened the pull request:
Postings always return impacts with freq=Integer.MAX_VALUE and norm=1 when frequencies are not indexed (IndexOptions.DOCS). This significantly overestimates the score upper bound of term queries, since the similarity scorer is effectively called with freq=1 all the time in this case (and either norm=1 if norms are not indexed, or the number of terms in the field otherwise).

This updates postings to always return impacts with freq=1 and norm=1 when frequencies are not indexed, which helps compute better score upper bounds, and in-turn makes dynamic pruning perform better.

Closes #14445

expani committed:
Ensuring skip list is read for fields indexed with only DOCS

Signed-off-by: expani <anijainc@amazon.com>

msfroh commented on lucene/core/src/java/org/apache/lucene/codecs/lucene103/Lucene103PostingsReader.java:
```diff
@@ -1310,7 +1317,7 @@ public List<Impact> getImpacts(int level) {
               }
             }
-            return DUMMY_IMPACTS;
+            return NON_COMPETITIVE_IMPACTS;
```
This was the only reference to `DUMMY_IMPACTS`, right? Can we remove it?

expani replied:
Good catch. 
This is unused after the change, so can remove it.

expani replied:
Looks like there was a case in CheckIndex failing due to this. 
So added it back for fields that have at least freqs enabled. 

Failing test
```
./gradlew :lucene:benchmark:test --tests "org.apache.lucene.benchmark.quality.TestQualityRun.testTrecQuality" -Ptests.jvms=1 -Ptests.jvmargs= -Ptests.seed=6AD36F6087951803 -Ptests.useSecurityManager=true -Ptests.gui=true -Ptests.file.encoding=ISO-8859-1 -Ptests.vectorsize=256 -Ptests.forceintegervectors=true
```

msfroh commented on lucene/core/src/java/org/apache/lucene/codecs/lucene103/Lucene103PostingsReader.java:
```diff
@@ -282,6 +288,10 @@ public PostingsEnum postings(
   @Override
   public ImpactsEnum impacts(FieldInfo fieldInfo, BlockTermState state, int flags)
       throws IOException {
+    if (state.docFreq <= BLOCK_SIZE) {
+      // no skip data
+      return new SlowImpactsEnum(postings(fieldInfo, state, null, flags));
+    }
```
This is essentially taking the place of `DUMMY_IMPACTS`, right? 

It's the thing that kicks in on tail blocks, which is what `DUMMY_IMPACTS` was there for. (I'm trying to make sure I understand the change.)

expani replied:
DUMMY_IMPACTS in PostingsReader was a `List<Impact>` 
whereas 
the DUMMY_IMPACTS returned by SlowImpactsEnum is a `Impacts` which would still apply for cases where using the skip list doesn't make sense `state.docFreq <= BLOCK_SIZE` 

I picked this check from `Lucene99PostingsReader` where we stored a default impact of freq=1 even if the field was indexed with `IndexOptions.DOCS`

expani replied:
You understanding is spot on.

The bug was this check getting extended to fields with `IndexOptions.DOCS`

expani committed:
refactor

Signed-off-by: expani <anijainc@amazon.com>

expani committed:
Fixed failing checkIndex tests

Signed-off-by: expani <anijainc@amazon.com>

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/lucene103/Lucene103PostingsReader.java:
```diff
@@ -282,6 +290,10 @@ public PostingsEnum postings(
+    if (state.docFreq <= BLOCK_SIZE) {
+      // no skip data
+      return new SlowImpactsEnum(postings(fieldInfo, state, null, flags));
+    }
```
Let's not use `SlowImpactsEnum` again or it would cancel the speedup from https://github.com/apache/lucene/pull/14017 (annotation HJ at https://benchmarks.mikemccandless.com/OrHighHigh.html).

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/lucene103/Lucene103PostingsReader.java:
```diff
@@ -66,13 +67,20 @@
+  // We stopped storing a placeholder impact with freq=1 for fields with IndexOptions.DOCS after
+  // 9.12.0
+  private static final List<Impact> NON_COMPETITIVE_IMPACTS =
+      Collections.singletonList(new Impact(1, 1L));
```
Maybe call it `DUMMY_IMPACTS_NO_FREQS` or something like that to draw a parallel with `DUMMY_IMPACTS` above?

expani replied:
Done

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/lucene103/Lucene103PostingsReader.java:
```diff
@@ -1286,14 +1298,11 @@ public long cost() {
           @Override
           public int numLevels() {
-            return indexHasFreq == false || level1LastDocID == NO_MORE_DOCS ? 1 : 2;
+            return level1LastDocID == NO_MORE_DOCS ? 1 : 2;
```
I would keep it the way it was for now. Exposing a single level that covers the whole doc ID space may cause inefficiencies with some scorers, but we should fix these scorers instead of tuning this class. If all docs have the same score, it should be just fine to return a single level of impact that covers the whole doc ID space.

expani replied:
I had made these changes to bring back the same behavior as 9.11.1

Without these changes, it doesn't read the skip data [ and exits here ](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/ImpactsDISI.java#L85) right after the min competitive score is set.

Whereas in 9.11.1, it reads the skip data [by entering this part of ImpactDISI](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/ImpactsDISI.java#L88-L97) which @msfroh described in the mail chain

```
It was fast because (once the collector has filled its priority queue), we'd check the (constant) impacts to find the first block that's strictly better than the min competitive score. Since all scores are equal, that would quickly skip to the end.
```

Although, not keeping this achieves the same result. Should I add a TODO to remove this later after we fix the scrorers you mentioned ? 

Also, can this give incorrect results when norms are enabled ? Since, we are not reading the impacts.

jpountz replied:
I understand why this change helps, but this problem is not unique to term queries indexed with IndexOptions.DOCS, `ConstantScoreScorer` (used by many queries) causes the same problem. I'd rather fix the root cause than merge this workaround that we may forget to remove later.

> Also, can this give incorrect results when norms are enabled ? Since, we are not reading the impacts.

Impacts help compute upper bounds of the score over ranges of doc IDs. Since scores are required to not increase when the norm increases, a score computed with norm=1 will always be greater than or equal to a score computed with any other norm value. So this is correct, it may just return a score upper bound that is greater than the actual best score from the block.

expani replied:
> Since scores are required to not increase when the norm increases

Wasn't aware of this. Makes sense now.

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/lucene103/Lucene103PostingsReader.java:
```diff
@@ -1286,14 +1298,11 @@ public long cost() {
           public int getDocIdUpTo(int level) {
-            if (indexHasFreq == false) {
-              return NO_MORE_DOCS;
-            }
```
Likewise, let's keep this?

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/lucene103/Lucene103PostingsReader.java:
```diff
@@ -1309,8 +1318,9 @@ public List<Impact> getImpacts(int level) {
+              return DUMMY_IMPACTS;
             }
-            return DUMMY_IMPACTS;
+            return NON_COMPETITIVE_IMPACTS;
```
I would keep it the way it was before and just add this at the beginning of this method (similar to `getDocIdUpTo`):

```java
            if (indexHasFreq == false) {
              return DUMMY_IMPACTS_NO_FREQS;
            }
```

expani replied:
Done

jpountz reviewed (commented):
Sorry for only seeing now, I left some suggestions as to how to update your change.

expani committed:
Addressed comments

Signed-off-by: expani <anijainc@amazon.com>

jpountz reviewed (approved):
LGTM. Can you add a CHANGES entry under 10.3 and undo the new line in SlowImpactsEnum?

expani committed:
Add changes entry

Signed-off-by: expani <anijainc@amazon.com>

expani committed:
Fixed formatting

Signed-off-by: expani <anijainc@amazon.com>

expani commented:
Addressed comments.
I want to backport these to `9.12.x` and `10.2.x` as well. Will open separate PRs for the same.

jpountz commented:
This sounds safe enough for 10.2.1 for me. Can you move the CHANGES entry to 10.2.1 then? cc @ChrisHegarty

jpountz commented:
I hope you don't mind, I updated this PR title and description to better reflect the change.

expani committed:
Moved change log to 10.2.1

Signed-off-by: expani <anijainc@amazon.com>

expani commented:
>I hope you don't mind, I updated this PR title and description to better reflect the change.

Not at all. Thanks for taking the time to explain the different pieces of this code. 

It was really fun debugging this and would definitely love to visit this part of the code again.

expani commented:
@ChrisHegarty @jpountz Moved the change log to 10.2.1

ChrisHegarty commented:
> @ChrisHegarty @jpountz Moved the change log to 10.2.1

Eh! I think you moved it to 10.2.0, rather than 10.2.1.

expani committed:
Rebased and moved to 10.2.1

Signed-off-by: expani <anijainc@amazon.com>

expani committed:
Rebased and moved to 10.2.1

Signed-off-by: expani <anijainc@amazon.com>

expani commented:
Oops hadn't rebased with main. Fixed it now.

ChrisHegarty commented:
> This sounds safe enough for 10.2.1 for me. Can you move the CHANGES entry to 10.2.1 then? cc @ChrisHegarty

What am I missing?  This is not applicable to 10.2.1, since the only changed file is Lucene103PostingsReader.java which is not present in 10.2 !

expani committed:
Same bug fix for 101PostingsReader

Signed-off-by: expani <anijainc@amazon.com>

expani committed:
Same bug fix for 101PostingsReader

Signed-off-by: expani <anijainc@amazon.com>

expani commented:
>What am I missing? This is not applicable to 10.2.1, since the only changed file is Lucene103PostingsReader.java which is not present in 10.2 ! Did the rebase mess something up ?

I had updated 103PostingsReader as the initial plan was not to backport. 

Updated 101PostingsReader which is used in 10.2.1 

Should I also raise against some other branch as well ?

expani committed:
Formatting

Signed-off-by: expani <anijainc@amazon.com>

expani commented:
I made the same change in `Lucene101PostingsReader` as well.

gf2121 commented:
To be clear, i raised https://github.com/apache/lucene/pull/14557 and https://github.com/apache/lucene/pull/14558 for backporting. I plan to merge this now if no one objects.

gf2121 commented:
@expani could you resolve the conflicts so that i can merge?

gf2121 committed:
Merge branch 'main' into perf_term_14445
